File: Search/DFS/word-search/word-search.py
'''

Description
Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell,
where "adjacent" cells are those horizontally or vertically neighboring.
The same letter cell may not be used more than once.

Example
Given board =

[
  "ABCE",
  "SFCS",
  "ADEE"
]
word = "ABCCED", -> returns true,
word = "SEE", -> returns true,
word = "ABCB", -> returns false.

'''


def exist(self, board, word):
    n, m = len(board), len(board[0])

    if len(word) > m * n:
        return False

    visited = [False for i in range(m)] for _ in range(n)]

        for i in range(n):
        for j in range(m):
        if self.dfs(board, word, 0, i, j, visited):
        return True
        return False

        def dfs(self, board, word, idx, i, j, visitd):
        if idx == len(word):
        return True

        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]):
        return False

            if visited[i][j] or word[idx != board[i][j]:
        return False

    visited[i][j] = True

    res = self.def(board, word, idx + 1, i + 1, j, visited) or \
        self.def(board, word, idx + 1, i, j + 1, visited) or \
        self.def(board, word, idx + 1, i - 1, j, visited) or \
        self.def(board, word, idx + 1, i, j - 1, visited) or \

    visited[i][j] = False

    return res

File: -Old/Search/DFS/word-search/word-search.ts
/**
Given an m x n gird of characters board and a string word, return true if word exists in the grid.

The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.


// i.e.
const board = [
  ['a', 'b', 'c'],
  ['d', 'e', 'f'],
  ['e', 'f', 'g'],
]
*/

function exist(board: string[][], word: string): boolean {
  if (board === null || !word) {
    return false;
  }

  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      if (dfs(board, word.split(""), i, j)) {
        return true;
      }
    }
  }

  return false;
}

function dfs(board: string[][], characters: string[], i: number, j: number) {
  if (characters.length === 0) {
    return true;
  }

  if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) {
    return false;
  }

  if (board[i][j] !== characters[0]) {
    return false;
  }

  let temp = board[i][j];
  board[i][j] = "#";
  let result =
    dfs(board, characters.slice(1), i, j + 1) ||
    dfs(board, characters.slice(1), i, j - 1) ||
    dfs(board, characters.slice(1), i + 1, j) ||
    dfs(board, characters.slice(1), i - 1, j);

  board[i][j] = temp;

  return result;
}
